clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -441,23 +441,26 @@ class radix_sort {
       unsigned_keys[i] = detail::traits<T>::twiddle_in(unsigned_keys[i]);
     }
 
-    while (true) {
+    for (int i = begin_bit; i <= end_bit; i += RADIX_BITS) {
       int pass_bits = sycl::min(RADIX_BITS, end_bit - begin_bit);
 
       int ranks[VALUES_PER_THREAD];
       detail::radix_rank<RADIX_BITS, DESCENDING>(_local_memory)
-          .template rank_keys(item, unsigned_keys, ranks, begin_bit, pass_bits);
-      begin_bit += RADIX_BITS;
+          .template rank_keys(item, unsigned_keys, ranks, i, pass_bits);
 
       item.barrier(sycl::access::fence_space::local_space);
 
+      if (is_striped && i == end_bit) {
+
+        exchange<T, VALUES_PER_THREAD>(_local_memory)
+            .scatter_to_striped(item, keys, ranks);
+
+        item.barrier(sycl::access::fence_space::local_space);
+      }
```

Comment:
An `else` here would be needed for correctness as we would falsely perform a `scatter_to_striped` followed by a `scatter_to_blocked` without it.